rfd/0207-vnet-ssh.md

```diff
@@ -0,0 +1,379 @@
+session certificate.
+Even though the generated OpenSSH config uses a ProxyCommand that calls `tsh` to
+dial the target SSH host, `tsh` can't handle per-session MFA here because the
+third-party SSH client is still ultimately responsible for authenticating to the
+host with the local SSH certificate from disk.
+
+Third-party SSH clients are unable to use
+[hardware keys](https://goteleport.com/docs/admin-guides/access-controls/guides/hardware-key-support/)
+because `tsh` uses a custom file format to reference these keys, which
+third-party clients won't recognize and won't be able to use for signatures.
+
+VNet SSH will automate all configuration steps to make SSH connections to
+Teleport servers simple and transparent while supporting advanced Teleport
+features that our users are accustomed to.
+
+## Details
+
+### UX
+
+When VNet SSH is active, third-party SSH clients such as OpenSSH will be able
+to connect to Teleport SSH hosts at a URI matching
+`<username>@<hostname>.<cluster name>`.
+For example, `ssh alice@node.teleport.example.com` will work to connect as
+the SSH user `alice` to a host named `node` in the cluster `teleport.example.com`.
+If per-session MFA is required, Connect will prompt the user to tap their
+hardware key.
+
+#### User story - VSCode remote development
+
+Alice is a software engineer with an underpowered laptop and she wants to
+compile and test her application on a more powerful machine.
+She starts a large EC2 instance in her company AWS account for development.
+Her company has an AMI that automatically joins the instance to their company
+Teleport cluster, `teleport.example.com`.
+The machine has the hostname `devbox`.
+
+Alice opens Teleport Connect and sees `devbox` as an SSH server on the main
+resources page.
+Next to the usual "Connect" dropdown on the SSH server card, she sees a new
+three-dots menu button.
+She clicks the menu button and there is a single option "Connect with VNet".
+She clicks the "Connect with VNet" button and a few things happen:
+
+1. Because Alice has not started VNet before, VNet starts for the first time.
+1. The VNet panel in the top left opens and displays:
+   > 🟢 Proxying TCP connections to teleport.example.com<br>
+   > 🔴 VNet SSH is not configured<br>
+   > Open Status Page
+1. A toast pops up in the bottom right and displays:
+   "VNet SSH is not configured, click here (link) to enable VNet SSH"
+
+Clicking either the "Open Status Page" button in the VNet panel or the link
+in the toast both open a new VNet Status tab in Connect.
+This tab will include the existing VNet diagnostic report, and a new section at
+the top on the status of VNet SSH.
+
+The SSH section in the VNet Status page will display:
+
+> VNet SSH lets any SSH client connect to Teleport SSH servers.<br>
+> When you enable VNet SSH, Connect will automatically add configuration options
+> to `~/.ssh/config`.
+>
+> Enable VNet SSH (button)
```

Comment:
@ravicious I'm just winging this let me know if you have thoughts 🙏

Reply:
The idea was to avoid editing the user's `~/.ssh/config` until they explicitly opt-in, I don't want to just do that by default. Do you think we have room to warn them about that before turning on VNet SSH in the VNet panel?

Reply:
We should be able to add a confirmation dialog for that.

Reply:
+1 for not touching `~/.ssh/config` without explicit permission.

A nice effect of using `Include` is that it's also pretty easy to detect if the line is still there, and to offer to add it again. It should also not be a requirement to do so to enable the server IMO, `ssh -F ~/path/to/vnet_ssh_config` will work just as well.

Reply:
The idea is everything in VNet will actually be enabled by default, `ssh -F ~/path/to/vnet_ssh_config` should work as Edoarda mentioned, Connect will just give you the option to automatically add the Include to `~/.ssh/config`

Reply:
One could think of presence of the `Include` line in `~/.ssh/config` as a sort of diagnostic, with an option to remedy that problem with a button on the diagnostic page. For the confirmation dialog you mentioned @ravicious I'm not sure when we'd open that